### PR TITLE
*: filter store with lots of pending peers

### PR DESCRIPTION
**Pull request by nolouch**
we need filter the store with lots of pending peers when scheduling regions.

### CODE REVIEW

File: server/cache.go [hunk 1]
```diff
@@ -341,6 +341,27 @@ func (c *clusterInfo) updateStoreStatus(id uint64) {
 	c.Stores.SetRegionCount(id, c.Regions.GetStoreRegionCount(id))
 }
 
+func (c *clusterInfo) updateStorePendingPeerCount(origin []*metapb.Peer, new []*metapb.Peer) {
+	m := make(map[uint64]int)
+	for _, o := range origin {
+		if num, ok := m[o.GetStoreId()]; ok {
```
Comment (siddontang) [inline server/cache.go]: seem you can use `m[o.GetStoreId()]--` directly.

**Comment by siddontang** [comment]
Rest LGTM

File: server/cache.go [hunk 2]
```diff
@@ -341,6 +341,27 @@ func (c *clusterInfo) updateStoreStatus(id uint64) {
 	c.Stores.SetRegionCount(id, c.Regions.GetStoreRegionCount(id))
 }
 
+func (c *clusterInfo) updateStorePendingPeerCount(origin []*metapb.Peer, new []*metapb.Peer) {
```
Comment (disksing) [inline server/cache.go]: I think you can use `core.RegionsInfo` to trace pending peers count of each store.

File: server/core/region.go [hunk 3]
```diff
@@ -333,6 +345,10 @@ func (r *RegionsInfo) RemoveRegion(region *RegionInfo) {
 		r.leaders[storeID].Delete(region.GetId())
 		r.followers[storeID].Delete(region.GetId())
 	}
+	for _, peer := range region.PendingPeers {
+		storeID := peer.GetStoreId()
+		r.pendingPeers[storeID].Delete(region.GetId())
```
Comment (disksing) [inline server/core/region.go]: Can we move it to L364? I think the pending peers are also included in peers list.

File: server/config.go [hunk 4]
```diff
@@ -305,7 +305,8 @@ func (c *Config) configFromFile(path string) error {
 type ScheduleConfig struct {
 	// If the snapshot count of one store is greater than this value,
 	// it will never be used as a source or target store.
-	MaxSnapshotCount uint64 `toml:"max-snapshot-count,omitempty" json:"max-snapshot-count"`
+	MaxSnapshotCount    uint64 `toml:"max-snapshot-count,omitempty" json:"max-snapshot-count"`
+	MaxPendingPeerCount uint64 `toml:"max-pending-peer-count,omitempty" json:"max-pending-peer-count"`
```
Comment (disksing) [inline server/config.go]: Need to add it to config file?

File: server/cache_test.go [hunk 5]
```diff
@@ -528,6 +528,53 @@ func (s *testClusterInfoSuite) testRegionSplitAndMerge(c *C, cache *clusterInfo)
...
+	newRegion := &core.RegionInfo{
+		Region:       &metapb.Region{Id: 1, Peers: peers[1:]},
+		Leader:       peers[1],
+		PendingPeers: peers[3:4],
+	}
+	cluster.handleRegionHeartbeat(newRegion)
+	checkPendingPeerCount([]int{0, 0, 0, 1}, cluster, c)
+}
+
+func checkPendingPeerCount(except []int, cache *clusterInfo, c *C) {
```
Comment (disksing) [inline server/cache_test.go]: except -> expect.

File: server/schedule/filters.go [hunk 6]
```diff
@@ -130,6 +130,31 @@ func (f *healthFilter) FilterTarget(store *core.StoreInfo) bool {
...
+}
+
+// NewPendingPeerCountFilter creates a Filter that filters all stores that are
+// currently handling too many pengding peers.
+func NewPendingPeerCountFilter(opt Options) Filter {
+	return &pendingPeerCountFilter{opt: opt}
+}
+
+func (p *pendingPeerCountFilter) filter(store *core.StoreInfo) bool {
+	if store != nil {
```
Comment (disksing) [inline server/schedule/filters.go]: I think it won't be nil.

**Pull request merged**